crates/yakui-core/src/dom/mod.rs

```diff
@@ -219,16 +226,24 @@ impl Dom {
 
             node.dynamic_scope_index = self.inner.dynamic_scope.current_scope();
             node.next_child = 0;
-            (id, widget)
+            (id, widget, node.callsite)
         };
 
         // Potentially recreate the widget, then update it.
         let response = {
-            if widget.as_ref().type_id() != TypeId::of::<T>() {
+            let is_type_different = widget.as_ref().type_id() != TypeId::of::<T>();
+            let is_callsite_different = callsite != Location::caller();
+
+            if is_type_different {
                 widget = Box::new(T::new());
             }
 
             let widget = widget.downcast_mut::<T>().unwrap();
+            // If the Type is different, it'd already be recreated. Let's avoid calling T::new twice.
+            if is_callsite_different && !is_type_different {
+                *widget = T::new();
+            }
```

Comment:
Can we merge these conditions?
```suggestion
            if is_type_different || is_callsite_different {
                widget = Box::new(T::new());
            }

            let widget = widget.downcast_mut::<T>().unwrap();
```

Reply:
Oh, good catch! The code was like that because we were considering making some function like `Widget::reset(&mut self);` instead that resets the state, but there doesn't seem to be a valid point for that right now.